allennlp/training/trainer.py

```diff
@@ -581,7 +581,12 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
             self._batch_num_total += 1
             batch_num_total = self._batch_num_total
 
-            self.optimizer.zero_grad()
+            # Zero gradients.
+            # NOTE: this is actually more efficient than calling `self.optimizer.zero_grad()`
+            # because it avoids a read op when the gradients are first updated below.
+            for param_group in self.optimizer.param_groups:
+                for p in param_group["params"]:
+                    p.grad = None
```

Comment:
Do the param groups necessarily cover all parameters?

Reply:
All parameters that require gradients, yes: https://github.com/allenai/allennlp/blob/master/allennlp/training/trainer.py#L1128-L1131